### ISSUE TITLE
Switching input pipes does not work anymore

### ISSUE THREAD

**Issue by tinbert** [ejurgensen/forked-daapd#824]
My installation in Stretch has several streams coming from gstreamer pipelines which are writing to distinct pipes.
If one pipe is selected in FD, the respective pipeline ist unpaused, music plays (on my AirPlays).
However, when I switch to the second pipe, the first gstreamer pipeline is stopped, but the second pipeline is not unpaused anymore, no music is playing.
This worked on Stretch for me, the problem occured after a fresh installation.

I have attached the debug log, unfortunately I can't really see which parts are relevant for tracking this bug, so please apologize for the maybe too lengthy log.

BTW: **Thx a lot! What a great product!!**

[forked-daapd.log](https://github.com/ejurgensen/forked-daapd/files/3674780/forked-daapd.log)

**Comment by ejurgensen** [comment]
Thanks! And the log snippet is just fine.

So I have made an attempt at reproducing the issue, but I didn't succeed with that. I also had a look at the log, and can't immediately see anything wrong. I see playback starting twice, first radioeins and then Klassikradio, which I suppose is the intention.

To help me reproduce the issue, can you send me detailed description of the steps? I am especially interested in what happens when you switch to the second pipe. When and how is writing to the first pipe stopped?

Another thing that I would like you to try is to try without Airplay, ie. by connecting headphones directly to you RPi. That is to find out if the issue has to do with Airplay not being restarted correctly.

**Comment by tinbert** [comment]
It looks like a problem with Airplay, now.

As you suggested, I turned off Airplay and used headphone jack. Then switching pipes was no proble!. With AirPlay turned on again, switching does not work, at least for the AirPlay destinations. However, on the headphone jack the sound from the new pipe was playing! 

I must mention, that switching MP3 files from the filesystem or internet streams is no problem, even with Airplay!

One note: My AirPlay installation consists of 3 Airport Express which I have not updated to Airplay 2, though it is possible. The problem occurs with single destinations, as well as with streaming to all 3 of the AEs.


The steps to reproduce: 
- restart RPi
- go to web interface on Mac ... :3689
- turn on 1 AirPlay destination (and "Computer" to test headphone jack)
- navigate to Files/music/Streams  (/home/pi/srv/music/Streams)
- Click on "radioeins"
-> Sound starts playing after some seconds
- Click on "Klassikradio"
-> Sound stops on AirPlay, no new sound on AirPlay
-> Sound Klassikradio starts on headphone jack, though!

The same problem occurs when using Remote app on iPhone/iPad.

**Comment by tinbert** [comment]
One followup: I found a workaround to switch pipes:
- play from one pipe to AirPlay
- stop playing pipe (press Stop)
- turn off AirPlay
- turn on AirPlay
- switch pipe -> plays!

**Comment by ejurgensen** [comment]
Ok, then it could be an issue with the timestamps being sent to the Airplay devices. Additional questions: I assume you have pipe_autostart disabled, since you are starting them manually in the steps you describe - can you confirm that?

**Comment by tinbert** [comment]
pipe_autostart false

Best Regards,
 
Robert

**Comment by ejurgensen** [comment]
I've tried as best I could to reproduce this issue, but it doesn't show up. I've tried with exactly the steps you mention, and both on my dev machine and on my RPi.

So to fix this I need you to do the bug-hunting. Your log does provide one clue, which is this line:
`Start sync packet sent to 'exp-office': cur_pos=432985977, cur_ts=40244:900699012, now=40247:170656525, rtptime=433073998`

"cur_ts" is the timestamp sent to the device, and "now" is the actual time. These should generally be close to each other, and indeed they are when I test. However, your log shows a difference of a 40247.17 - 40244.90 = 2.27 seconds. I think that if the difference becomes over 2 seconds then some devices will not start playback, because then the timing doesn't make sense.

So the first thing you would need to test is if there is anything to this theory. I.e. that if the difference is high, then there is no playback, and if it is low there is playback. If it turns out to be the case then the next step is to start finding out what is causing the timestamp difference. I can help you with that at that point.
